dpnp/backend/extensions/sycl_ext/histogram.cpp

```diff
@@ -0,0 +1,266 @@
+        dpctl_td_ns::TypePairDefinedEntry<dT, float, hT, std::complex<float>>,
+        dpctl_td_ns::TypePairDefinedEntry<dT, double, hT, std::complex<double>>,
+        dpctl_td_ns::TypePairDefinedEntry<dT, std::complex<float>, hT, int64_t>,
+        dpctl_td_ns::
+            TypePairDefinedEntry<dT, std::complex<double>, hT, int64_t>,
+        dpctl_td_ns::TypePairDefinedEntry<dT, std::complex<float>, hT, float>,
+        dpctl_td_ns::TypePairDefinedEntry<dT, std::complex<double>, hT, double>,
+        // fall-through
+        dpctl_td_ns::NotDefinedEntry>::is_defined;
+
+    fnT get()
+    {
+        if constexpr (is_defined) {
+            return histogram_impl<dT, dT, hT>;
+        }
+        else {
+            return nullptr;
+        }
+    }
+};
+
+using sycl_ext::histogram::Histogram;
+
+Histogram::FnT
+    dispatch(Histogram *hist, int data_typenum, int, int hist_typenum)
+{
+    auto array_types = dpctl_td_ns::usm_ndarray_types();
+    const int data_type_id = array_types.typenum_to_lookup_id(data_typenum);
+    const int hist_type_id = array_types.typenum_to_lookup_id(hist_typenum);
+
+    auto histogram_fn = hist->dispatch_table[data_type_id][hist_type_id];
+
+    if (histogram_fn == nullptr) {
+        throw py::value_error("Unsupported data types"); // report types?
+    }
+
+    return histogram_fn;
+}
+
+} // namespace
+
+Histogram::Histogram()
+{
+    dpctl_td_ns::DispatchTableBuilder<FnT, ContigFactory,
+                                      dpctl_td_ns::num_types>
+        contig;
+    contig.populate_dispatch_table(dispatch_table);
+}
+
+std::tuple<sycl::event, sycl::event>
+    Histogram::call(const dpctl::tensor::usm_ndarray &sample,
+                    const dpctl::tensor::usm_ndarray &bins,
+                    std::optional<const dpctl::tensor::usm_ndarray> &weights,
+                    dpctl::tensor::usm_ndarray &histogram,
+                    const std::vector<sycl::event> &depends)
+{
+    validate(sample, bins, weights, histogram);
+
+    const int sample_typenum = sample.get_typenum();
+    const int bins_typenum = bins.get_typenum();
+    const int hist_typenum = histogram.get_typenum();
+
+    auto histogram_func =
+        dispatch(this, sample_typenum, bins_typenum, hist_typenum);
+
+    auto exec_q = sample.get_queue();
+
+    void *weights_ptr =
+        weights.has_value() ? weights.value().get_data() : nullptr;
+
+    auto ev =
+        histogram_func(exec_q, sample.get_data(), bins.get_data(), weights_ptr,
+                       histogram.get_data(), histogram.get_shape(0),
+                       sample.get_shape(0), depends);
+
+    sycl::event args_ev;
+    if (weights.has_value()) {
+        args_ev = dpctl::utils::keep_args_alive(
+            exec_q, {sample, bins, weights.value(), histogram}, {ev});
+    }
+    else {
+        args_ev = dpctl::utils::keep_args_alive(
+            exec_q, {sample, bins, histogram}, {ev});
+    }
+
+    return {ev, args_ev};
```

Comment:
In other extensions we firstly return a host event. I'd propose to keep consistency here:
```suggestion
    return {args_ev, ev};
```

Reply:
Thanks! I've changed the order